File: src/main/java/org/tindalos/principle/domain/detector/structure/PackageStructureHints1Finder.scala
package org.tindalos.principle.domain.detector.structure

import org.tindalos.principle.domain.detector.structure.Graph._

object PackageStructureHints1Finder {

  case class GroupingResult(val grouping: Map[Set[String], List[NodeId]], val labelledSources: List[(String, NodeId)])

  def makeGroups(graph: Set[Node]) = {
    val sources = Graph.findSources(graph).toList.sortBy(_.id)
    val labelledSources = for {i <- 0 to sources.size - 1} yield (sources(i), label(sources.size, i))

    val labelledNodes: IndexedSeq[(String, NodeId)] =
      for {(source, label) <- labelledSources
           downstream <- Graph.findDownstreamNodes(source, graph)} yield (label, downstream.id)

    val x1: Map[NodeId, List[(String, NodeId)]] = labelledNodes
      .toList
      .groupBy(_._2)

    val x2: Map[NodeId, Set[String]] = x1.map(kv => (kv._1, kv._2.map(_._1).toSet))

    val g = x2.toList.groupBy(_._2).map(kv => (kv._1, kv._2.map(_._1)))

    GroupingResult(g, labelledSources.toList.map(x => (x._2, x._1.id)))

  }

  def label(max: Int, i: Int) = {
    "".r
    val postfix =
      if (max < 10) i.toString
      else if (max < 100)
        if (i < 10) "0" + i
        else i.toString
      else if (max < 1000)
        if (i < 10) "00" + i
        else if (i < 100) "0" + i
        else i.toString
    "s" + postfix
  }

}

File: src/main/java/org/tindalos/principle/internal/domain/analyzers/structure/PackageStructureHints1Finder.java
package org.tindalos.principle.internal.domain.analyzers.structure;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.tindalos.principle.internal.domain.analyzers.structure.GroupingResult.LabelledSource;
import org.tindalos.principle.internal.domain.core.Node;

/**
 * Creates source-based grouping hints from a class dependency graph.
 */
public final class PackageStructureHints1Finder {

    private PackageStructureHints1Finder() {
    }

    public static GroupingResult makeGroups(Set<Node> graph) {
        var sources = Graph.findSources(graph).stream()
                .sorted(Comparator.comparing(Node::id))
                .toList();

        var labelledSources = new ArrayList<Map.Entry<Node, String>>();
        for (int i = 0; i < sources.size(); i++) {
            labelledSources.add(Map.entry(sources.get(i), label(sources.size(), i)));
        }

        var groupedLabelsByNode = new HashMap<String, Set<String>>();
        for (var labelledSource : labelledSources) {
            var source = labelledSource.getKey();
            var sourceLabel = labelledSource.getValue();
            for (var downstream : Graph.findDownstreamNodes(source, graph)) {
                groupedLabelsByNode.merge(
                        downstream.id(),
                        Set.of(sourceLabel),
                        (existing, added) -> {
                            var merged = new java.util.HashSet<>(existing);
                            merged.addAll(added);
                            return Set.copyOf(merged);
                        });
            }
        }

        var groupedNodeIdsBySourceSet = new LinkedHashMap<Set<String>, List<String>>();
        for (var entry : groupedLabelsByNode.entrySet()) {
            groupedNodeIdsBySourceSet
                    .computeIfAbsent(entry.getValue(), ignored -> new ArrayList<>())
                    .add(entry.getKey());
        }

        var javaGrouping = groupedNodeIdsBySourceSet.entrySet().stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        entry -> List.copyOf(entry.getValue()),
                        (left, right) -> left,
                        LinkedHashMap::new));

        var javaLabelledSources = labelledSources.stream()
                .map(entry -> new LabelledSource(entry.getValue(), entry.getKey().id()))
                .toList();

        return new GroupingResult(javaGrouping, javaLabelledSources);
    }

    public static String label(int max, int i) {
        String postfix;
        if (max < 10) {
            postfix = Integer.toString(i);
        } else if (max < 100) {
            postfix = i < 10 ? "0" + i : Integer.toString(i);
        } else if (max < 1000) {
            if (i < 10) {
                postfix = "00" + i;
            } else {
                postfix = i < 100 ? "0" + i : Integer.toString(i);
            }
        } else {
            postfix = Integer.toString(i);
        }
        return "s" + postfix;
    }
}